Update StringView::UpconvertedCharactersWithSize to store its size

cdumez opened the pull request:
#### fd9e18c7f3c6303f8a7bb1d8b0af8767bb392410
<pre>
Update StringView::UpconvertedCharactersWithSize to store its size
<a href="https://bugs.webkit.org/show_bug.cgi?id=272629">https://bugs.webkit.org/show_bug.cgi?id=272629</a>

Reviewed by Darin Adler.

Update StringView::UpconvertedCharactersWithSize to store its size and add a
span() function.

* Source/JavaScriptCore/runtime/IntlCache.cpp:
(JSC::IntlCache::getBestDateTimePattern):
* Source/JavaScriptCore/runtime/IntlCache.h:
* Source/JavaScriptCore/runtime/IntlDateTimeFormat.cpp:
(JSC::IntlDateTimeFormat::initializeDateTimeFormat):
* Source/WTF/wtf/text/StringImpl.cpp:
(WTF::StringImpl::convertToUppercaseWithoutLocale):
(WTF::StringImpl::convertToLowercaseWithLocale):
(WTF::StringImpl::convertToUppercaseWithLocale):
(WTF::StringImpl::foldCase):
* Source/WTF/wtf/text/StringView.h:
(WTF::StringView::UpconvertedCharactersWithSize::operator const UChar* const):
(WTF::StringView::UpconvertedCharactersWithSize::get const):
(WTF::StringView::UpconvertedCharactersWithSize::span const):
(WTF::StringView::UpconvertedCharactersWithSize&lt;N&gt;::UpconvertedCharactersWithSize):
* Source/WebCore/PAL/pal/text/TextCodecCJK.cpp:
(PAL::eucJPEncode):
(PAL::iso2022JPEncode):
(PAL::shiftJISEncode):
(PAL::eucKREncode):
(PAL::big5Encode):
(PAL::gbEncodeShared):
* Source/WebCore/editing/Editor.cpp:
(WebCore::scanForTelephoneNumbers):
* Source/WebCore/html/parser/HTMLTreeBuilder.cpp:
(WebCore::HTMLTreeBuilder::linkifyPhoneNumbers):
* Source/WebCore/platform/Length.cpp:
(WebCore::newLengthArray):
* Source/WebCore/platform/SharedStringHash.cpp:
(WebCore::computeVisitedLinkHash):
* Source/WebCore/platform/TelephoneNumberDetector.h:
* Source/WebCore/platform/cocoa/TelephoneNumberDetectorCocoa.cpp:
(WebCore::TelephoneNumberDetector::find):
* Source/WebCore/platform/graphics/StringTruncator.cpp:
(WebCore::stringWidth):
(WebCore::truncateString):
(WebCore::StringTruncator::width):
* Tools/TestWebKitAPI/Tests/WTF/StringView.cpp:
(TestWebKitAPI::TEST(WTF, StringViewUpconvert)):

Canonical link: <a href="https://commits.webkit.org/277464@main">https://commits.webkit.org/277464@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8742d8f6b7da579ae9a9687883351af0717ed195

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/47685 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/26873 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/50436 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/50368 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/43740 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/49992 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/32726 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/24351 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/38818 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/48267 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/24532 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/41163 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/20115 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/22013 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/42354 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/5734 "Built successfully") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/40967 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/44034 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/42776 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/52260 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/47176 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/22721 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/19055 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/46123 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/23993 "Built successfully") | | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/45154 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/24781 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/54676 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6742 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/23713 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/11219 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WTF/wtf/text/StringView.h:
```diff
@@ -521,11 +521,13 @@ class StringView::UpconvertedCharactersWithSize {
-    const UChar* get() const { return m_characters; }
+    operator const UChar*() const { return m_characters.data(); }
+    const UChar* get() const { return m_characters.data(); }
+    std::span<const UChar> span() const { return m_characters; }
```
We could also add an `operator std::span<const UChar>` for the same reason we added an `operator const UChar*`. This would allow us to omit the `span()` at most call sites. That’s sort of the "trick" of this class’s original design. We can sort of pretend it is just returning a pointer to characters that already exist because the lifetime comes out correctly when you call `upconvertedCharacters` in a function argument expression.

Or you can leave it out. We don’t need to be tricky, I guess.

darinadler commented on Source/WebCore/platform/Length.cpp:
```diff
@@ -105,15 +105,15 @@ UniqueArray<Length> newLengthArray(const String& string, int& len)
-    if (str->length()-pos > 0)
-        r[i] = parseLength({ upconvertedCharacters + pos, str->length() - pos });
+    if (str->length() - pos > 0)
+        r[i] = parseLength(upconvertedCharacters.span().subspan(pos, str->length() - pos));
```
Don’t need the `str->length - pos` here because the span already knows the string’s length.

darinadler commented on Source/WebCore/html/parser/HTMLTreeBuilder.cpp:
```diff
@@ -2451,33 +2451,35 @@ void HTMLTreeBuilder::linkifyPhoneNumbers(const String& string)
-        if (scannerPosition < length) {
-            String after = string.substring(scannerPosition, length - scannerPosition);
+        if (scannerPosition < string.length()) {
+            String after = string.substring(scannerPosition);
```
Seems like it would be nicer without the local variable.

cdumez committed:
Update StringView::UpconvertedCharactersWithSize to store its size
https://bugs.webkit.org/show_bug.cgi?id=272629

Reviewed by Darin Adler.

Update StringView::UpconvertedCharactersWithSize to store its size and add a
span() function.

* Source/JavaScriptCore/runtime/IntlCache.cpp:
(JSC::IntlCache::getBestDateTimePattern):
* Source/JavaScriptCore/runtime/IntlCache.h:
* Source/JavaScriptCore/runtime/IntlDateTimeFormat.cpp:
(JSC::IntlDateTimeFormat::initializeDateTimeFormat):
* Source/WTF/wtf/text/StringImpl.cpp:
(WTF::StringImpl::convertToUppercaseWithoutLocale):
(WTF::StringImpl::convertToLowercaseWithLocale):
(WTF::StringImpl::convertToUppercaseWithLocale):
(WTF::StringImpl::foldCase):
* Source/WTF/wtf/text/StringView.h:
(WTF::StringView::UpconvertedCharactersWithSize::operator const UChar* const):
(WTF::StringView::UpconvertedCharactersWithSize::get const):
(WTF::StringView::UpconvertedCharactersWithSize::span const):
(WTF::StringView::UpconvertedCharactersWithSize<N>::UpconvertedCharactersWithSize):
* Source/WebCore/PAL/pal/text/TextCodecCJK.cpp:
(PAL::eucJPEncode):
(PAL::iso2022JPEncode):
(PAL::shiftJISEncode):
(PAL::eucKREncode):
(PAL::big5Encode):
(PAL::gbEncodeShared):
* Source/WebCore/editing/Editor.cpp:
(WebCore::scanForTelephoneNumbers):
* Source/WebCore/html/parser/HTMLTreeBuilder.cpp:
(WebCore::HTMLTreeBuilder::linkifyPhoneNumbers):
* Source/WebCore/platform/Length.cpp:
(WebCore::newLengthArray):
* Source/WebCore/platform/SharedStringHash.cpp:
(WebCore::computeVisitedLinkHash):
* Source/WebCore/platform/TelephoneNumberDetector.h:
* Source/WebCore/platform/cocoa/TelephoneNumberDetectorCocoa.cpp:
(WebCore::TelephoneNumberDetector::find):
* Source/WebCore/platform/graphics/StringTruncator.cpp:
(WebCore::stringWidth):
(WebCore::truncateString):
(WebCore::StringTruncator::width):
* Tools/TestWebKitAPI/Tests/WTF/StringView.cpp:
(TestWebKitAPI::TEST(WTF, StringViewUpconvert)):

Canonical link: https://commits.webkit.org/277464@main
